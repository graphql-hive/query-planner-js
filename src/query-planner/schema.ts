import { JoinField, JoinType } from "./join-directives";

export enum TypeKind {
  OBJECT_TYPE,
  INTERFACE_TYPE,
  ENUM_TYPE,
  UNION_TYPE,
  INPUT_OBJECT_TYPE,
  SCALAR_TYPE,
}

export class Supergraph {
  public subgraphs = new Map<string, Subgraph>();
}

export class Subgraph {
  graphId: string;
  types: Map<string, ObjectType> = new Map();
  entityTypes = new Set<string>();

  constructor(graphId: string) {
    this.graphId = graphId;
  }
}

export class ObjectType {
  public name: string;
  public fields: ObjectTypeField[] = [];
  public join: JoinType[] = [];

  constructor(name: string, join: JoinType[]) {
    this.name = name;
    this.join = join;
  }

  addField(field: ObjectTypeField) {
    this.fields.push(field);
  }
}

export class ObjectTypeField {
  public name: string;
  public type: string;
  public isList: boolean;
  public join: JoinField;

  constructor(name: string, type: string, isList: boolean, join: JoinField) {
    this.name = name;
    this.type = type;
    this.isList = isList;
    this.join = join;
  }
}
