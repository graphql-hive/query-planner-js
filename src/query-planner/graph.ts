import { TypeKind } from "./schema";
import { SelectionResolver } from "./selection-resolver";
import { invariant } from "./utils";

interface Display {
  toString(): string;
}

export class Node implements Display {
  constructor(
    /**
     * Unique index in the graph
     */
    public index: number,
    public subgraphId: string,
    public typeName: string,
    public typeKind: TypeKind,
  ) {}

  toString() {
    return `${this.typeName}/${this.subgraphId}`;
  }
}

type Move = (FieldMove | EntityMove | AbstractMove | InterfaceObjectMove) &
  Display;

export class FieldMove implements Display {
  constructor(
    public fieldName: string,
    public typeName: string,
    public typeKind: TypeKind,
  ) {}

  toString() {
    return `${this.fieldName}`;
  }
}

export class EntityMove implements Display {
  toString() {
    return "ðŸ”‘";
  }
}

export class AbstractMove implements Display {
  constructor(
    public fromTypeName: string,
    public fromTypeKind: TypeKind,
    public toTypeName: string,
    public toTypeKind: TypeKind,
  ) {}

  toString() {
    return `... on ${this.toTypeName}`;
  }
}

export class InterfaceObjectMove implements Display {
  constructor(
    public fromTypeName: string,
    public fromTypeKind: TypeKind,
    public toTypeName: string,
  ) {}

  toString() {
    return `... on ${this.toTypeName}`;
  }
}

export class Edge<T = Move> implements Display {
  constructor(
    public head: Node,
    public tail: Node,
    public move: T,
    public requirement: Selection | null,
  ) {}

  toString(): string {
    return `${this.head.toString()} -(${(this.move as Display).toString()} ${this.requirement?.toString() ?? ""})- ${this.tail.toString()}`;
  }
}

export class Graph {
  public id: string;
  // We do it for two reasons:
  // 1. We want to be able to quickly find all nodes/edges of a given type
  // 2. We want to avoid array length limit
  private nodesByTypeIndex: Node[][] = [];
  // We have two indexes of edges:
  // 1. By head type
  // 2. By tail type
  // We do it to quickly pick edges by head/tail type, without iterating over all edges.
  private edgesByHeadTypeIndex: Edge[][] = [];
  private edgesByTailTypeIndex: Edge[][] = [];
  // To quickly find all nodes of a given type
  private typeNameToNodeIndexes = new Map<string, number[]>();

  constructor(id: string) {
    this.id = id;
  }

  ensureNonOrSingleNode(typeName: string) {
    const indexes = this.typeNameToNodeIndexes.get(typeName);

    if (!Array.isArray(indexes)) {
      return;
    }

    if (indexes.length > 1) {
      throw new Error(
        `Expected only one node for ${typeName} in graph ${this.id}`,
      );
    }

    return this.nodesByTypeIndex[indexes[0]][0];
  }

  createTypeNode(typeName: string, typeKind: TypeKind) {
    if (this.typeNameToNodeIndexes.has(typeName)) {
      throw new Error(
        `Node for ${typeName} already exists in subgraph ${this.id}`,
      );
    }

    return this.createNode(this.id, typeName, typeKind);
  }

  private createNode(graphId: string, typeName: string, typeKind: TypeKind) {
    const index = this.nodesByTypeIndex.push([]) - 1;
    const node = new Node(index, graphId, typeName, typeKind);
    this.nodesByTypeIndex[node.index].push(node);
    this.edgesByHeadTypeIndex.push([]);
    this.edgesByTailTypeIndex.push([]);

    const existing = this.typeNameToNodeIndexes.get(typeName);

    if (Array.isArray(existing)) {
      existing.push(index);
    } else {
      this.typeNameToNodeIndexes.set(typeName, [index]);
    }

    return node;
  }

  addEdge(edge: Edge) {
    const edgeIndex = this.edgesByHeadTypeIndex[edge.head.index].push(edge) - 1;
    this.edgesByTailTypeIndex[edge.tail.index].push(edge);
    return edge;
  }

  addNode(node: Node) {
    const newIndex = this.nodesByTypeIndex.push([]) - 1;
    node.index = newIndex;

    this.nodesByTypeIndex[node.index].push(node);
    this.edgesByHeadTypeIndex.push([]);
    this.edgesByTailTypeIndex.push([]);

    const existing = this.typeNameToNodeIndexes.get(node.typeName);

    if (Array.isArray(existing)) {
      existing.push(newIndex);
    } else {
      this.typeNameToNodeIndexes.set(node.typeName, [newIndex]);
    }

    return node;
  }

  copyFrom(graph: Graph) {
    for (const nodes of graph.nodesByTypeIndex) {
      for (const node of nodes) {
        this.addNode(node);
      }
    }

    for (const edges of graph.edgesByHeadTypeIndex) {
      for (const edge of edges) {
        this.addEdge(edge);
      }
    }
  }

  joinByKeys(
    entities: Map<
      string,
      Array<{
        key: string;
        graphId: string;
        typeName: string;
      }>
    >,
    selectionResolvers: Map<string, SelectionResolver>,
  ) {
    // for each entity type, we want to assign one entity node to a matching entity node in other subgraphs
    const edgesToAdd: Edge[] = [];

    for (let i = 0; i < this.nodesByTypeIndex.length; i++) {
      const typeNode = this.nodesByTypeIndex[i][0];
      const entitiesOfType = entities.get(typeNode.typeName);

      if (!Array.isArray(entitiesOfType)) {
        continue; // no entities of this type
      }

      const otherNodesIndexes = this.typeNameToNodeIndexes.get(
        typeNode.typeName,
      );

      if (!Array.isArray(otherNodesIndexes)) {
        continue;
      }

      if (typeNode.typeKind !== TypeKind.OBJECT_TYPE) {
        // We will only support objects for now
        continue;
      }

      this.connectEntities(
        i,
        otherNodesIndexes,
        edgesToAdd,
        entitiesOfType,
        selectionResolvers,
      );
    }

    while (edgesToAdd.length > 0) {
      const edge = edgesToAdd.pop();

      if (!edge) {
        throw new Error("Expected edge to be defined");
      }

      this.addEdge(edge);
    }
  }

  print(asLink = false) {
    let str = "digraph G {";

    if (this.typeNameToNodeIndexes.has("Query")) {
      str += "\n root -> Query";
    }

    for (const edge of this.edgesByHeadTypeIndex.flat()) {
      if (edge.head.typeName === "Query") {
        str += `\n  "Query" -> "${edge.head}";`;
      } else if (edge.head.typeName === "Mutation") {
        str += `\n  "Mutation" -> "${edge.head}";`;
      } else if (edge.head.typeName === "Subscription") {
        str += `\n  "Subscription" -> "${edge.head}";`;
      }

      const label = edge.move.toString() + " " + edge.requirement?.toString();
      str += `\n  "${edge.head}" -> "${edge.tail}" [label="${label}"];`;
    }

    str += "\n}";

    if (asLink) {
      return `https://dreampuf.github.io/GraphvizOnline/#${encodeURIComponent(str)}`;
    }

    return str;
  }

  private connectEntities(
    nodeIndex: number,
    sameTypeNameNodeIndexes: number[],
    edgesToAdd: Edge[],
    entitiesOfType: Array<{
      key: string;
      graphId: string;
      typeName: string;
    }>,
    selectionResolvers: Map<string, SelectionResolver>,
  ) {
    for (const headNode of this.nodesByTypeIndex[nodeIndex]) {
      for (const otherNodeIndex of sameTypeNameNodeIndexes) {
        if (nodeIndex === otherNodeIndex) {
          continue;
        }

        for (const tailNode of this.nodesByTypeIndex[otherNodeIndex]) {
          if (!entitiesOfType.some((t) => t.graphId === tailNode.subgraphId)) {
            continue;
          }

          for (const { key } of entitiesOfType.filter(
            (t) => t.graphId === tailNode.subgraphId,
          )) {
            const selectionResolver = selectionResolvers.get(
              tailNode.subgraphId,
            );
            invariant(selectionResolver, "keyFieldsResolver is not defined");

            edgesToAdd.push(
              new Edge(
                headNode,
                tailNode,
                new EntityMove(),
                selectionResolver.resolve(headNode.typeName, key),
              ),
            );
          }
        }
      }
    }
  }

  private getIndexesOfType(typeName: string) {
    return this.typeNameToNodeIndexes.get(typeName);
  }

  nodesOf(typeName: string, failIfMissing = true) {
    const indexes = this.getIndexesOfType(typeName);

    if (!Array.isArray(indexes)) {
      if (failIfMissing) {
        throw new Error(
          `Expected TypeNode(${typeName}) to be inserted first in graph ${this.id}`,
        );
      }

      return [];
    }

    const nodes: Node[] = [];

    for (const i of indexes) {
      for (const node of this.nodesByTypeIndex[i]) {
        nodes.push(node);
      }
    }

    return nodes;
  }

  nodeOf(typeName: string): Node;
  nodeOf(typeName: string, failIfMissing: true): Node;
  nodeOf(typeName: string, failIfMissing?: false): Node | undefined;
  nodeOf(typeName: string, failIfMissing = true) {
    const indexes = this.getIndexesOfType(typeName);

    if (!Array.isArray(indexes)) {
      if (failIfMissing) {
        throw new Error(
          `Expected TypeNode(${typeName}) to be inserted first in graph ${this.id}`,
        );
      }

      return undefined;
    }

    if (indexes.length > 1) {
      throw new Error(
        `Expected only one node for ${typeName} in graph ${this.id}`,
      );
    }

    return this.nodesByTypeIndex[indexes[0]][0];
  }

  edgesOfHead(head: Node) {
    const filtered = this.edgesByHeadTypeIndex[head.index]?.filter(
      (e) => e.head === head,
    );

    if (filtered.length !== this.edgesByHeadTypeIndex[head.index].length) {
      throw new Error("poop");
    }

    return this.edgesByHeadTypeIndex[head.index];
  }
}

export type Field = {
  kind: "field";
  typeName: string;
  fieldName: string;
  selectionSet: null | Array<SelectionNode>;
};

export type Fragment = {
  kind: "fragment";
  typeName: string;
  selectionSet: Array<SelectionNode>;
};

export type SelectionNode = Field | Fragment;

export class Selection {
  constructor(
    private typeName: string,
    private keyFieldsString: string,
    public selectionSet: SelectionNode[],
  ) {}

  contains(typeName: string, fieldName: string) {
    return this._contains(typeName, fieldName, this.selectionSet);
  }

  equals(other: Selection) {
    if (this.typeName !== other.typeName) {
      return false;
    }

    if (this.keyFieldsString === other.keyFieldsString) {
      return true;
    }

    return this._selectionSetEqual(this.selectionSet, other.selectionSet);
  }

  private _selectionSetEqual(
    selectionSet: SelectionNode[],
    otherSelectionSet: SelectionNode[],
  ): boolean {
    if (selectionSet.length !== otherSelectionSet.length) {
      return false;
    }

    for (let i = 0; i < selectionSet.length; i++) {
      // Fields are sorted by typeName and fieldName, so we can compare them directly.
      // See: SelectionResolver#sort
      const selectionNode = selectionSet[i];
      const otherSelectionNode = otherSelectionSet[i];

      if (selectionNode.kind !== otherSelectionNode.kind) {
        return false;
      }

      // Compare typeName and fieldName
      if (selectionNode.typeName !== otherSelectionNode.typeName) {
        return false;
      }

      if (
        selectionNode.kind === "field" &&
        otherSelectionNode.kind === "field" &&
        selectionNode.fieldName !== otherSelectionNode.fieldName
      ) {
        return false;
      }

      const areEqual =
        // Compare selectionSet if both are arrays
        // Otherwise, compare nullability of selectionSet
        Array.isArray(selectionNode.selectionSet) &&
        Array.isArray(otherSelectionNode.selectionSet)
          ? this._selectionSetEqual(
              selectionNode.selectionSet,
              otherSelectionNode.selectionSet,
            )
          : selectionNode.selectionSet === otherSelectionNode.selectionSet;

      // Avoid unnecessary iterations if we already know that fields are not equal
      if (!areEqual) {
        return false;
      }
    }

    return true;
  }

  private _contains(
    typeName: string,
    fieldName: string,
    selectionSet: SelectionNode[],
  ): boolean {
    return selectionSet.some(
      (s) =>
        (s.kind === "field" &&
          s.typeName === typeName &&
          s.fieldName === fieldName) ||
        (s.selectionSet
          ? this._contains(typeName, fieldName, s.selectionSet)
          : false),
    );
  }

  toString() {
    return this.keyFieldsString;
  }
}
