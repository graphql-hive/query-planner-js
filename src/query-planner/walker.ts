import {
  AbstractMove,
  Edge,
  EntityMove,
  FieldMove,
  Graph,
  InterfaceObjectMove,
  Node,
  Selection,
} from "./graph.js";
import { OperationTypeNode } from "graphql";
import { invariant } from "./utils.js";

type Step = FieldStep;

type FieldStep = {
  kind: "Field";
  /**
   * Name of the field, we want to want to walk to
   */
  name: string;
};

export function walkQuery(
  graph: Graph,
  operationType: OperationTypeNode,
  steps: Step[],
) {
  const initialPaths = graph
    .nodesOf(operationTypeToTypeName(operationType))
    .map((node) => new OperationPath(node));
  const stepsCloned = steps.slice();
  const stack: [Step, OperationPath[]][] = [
    [stepsCloned.shift()!, initialPaths],
  ];

  while (stack.length > 0) {
    const [step, paths] = stack.pop()!;

    console.log("Processing step:", step);

    if (step.kind !== "Field") {
      throw new Error(`Unsupported step kind: ${(step as any).kind}`);
    }

    const nextPaths: OperationPath[] = [];

    for (const path of paths) {
      const directPaths = findDirectPaths(graph, path, step);

      if (directPaths.length) {
        nextPaths.push(...directPaths);
      }

      const indirectPaths = findIndirectPaths(graph, path, step);
      if (indirectPaths.length) {
        nextPaths.push(...indirectPaths);
      }
    }

    const nextStep = stepsCloned.shift();

    if (nextPaths.length && nextStep) {
      console.log("Found", nextPaths.length, "paths for", step);
      stack.push([nextStep, nextPaths]);
    }

    if (!nextStep) {
      return nextPaths;
    }
  }

  // find best path

  return [];
}

function operationTypeToTypeName(operationType: OperationTypeNode): string {
  switch (operationType) {
    case OperationTypeNode.QUERY:
      return "Query";
    case OperationTypeNode.MUTATION:
      return "Mutation";
    case OperationTypeNode.SUBSCRIPTION:
      return "Subscription";
    default:
      throw new Error(`Unsupported operation type: ${operationType}`);
  }
}

function findDirectPaths(
  graph: Graph,
  path: OperationPath,
  step: Step,
): OperationPath[] {
  const nextPaths: OperationPath[] = [];
  const edges = graph.edgesOfHead(path.tail());

  if (step.kind !== "Field") {
    // TODO: we do not support non-field steps yet
    return nextPaths;
  }

  for (const edge of edges) {
    if (!(edge.move instanceof FieldMove)) {
      continue;
    }

    if (edge.move.fieldName !== step.name) {
      continue;
    }

    if (edge.move.fieldName === step.name) {
      nextPaths.push(path.advance(edge));
    }
  }

  return nextPaths;
}

function findIndirectPaths(
  graph: Graph,
  path: OperationPath,
  step: Step,
): OperationPath[] {
  const nextPaths: OperationPath[] = [];
  const tail = path.tail();
  const sourceGraphId = tail.subgraphId;

  const queue: [string[], Selection[], OperationPath][] = [[[], [], path]];

  while (queue.length > 0) {
    const item = queue.pop();
    invariant(!!item, "Unexpected end of queue");

    const [visitedGraphs, visitedKeyFields, path] = item;
    const tail = path.tail();
    const edges = graph.edgesOfHead(tail);

    for (const edge of edges) {
      if (visitedGraphs.includes(edge.tail.subgraphId)) {
        // Already visited graph
        continue;
      }

      if (edge.tail.subgraphId === sourceGraphId) {
        // Prevent a situation where we are going back to the same graph
        // The only exception is when we are moving to an abstract type
        continue;
      }

      if (!(edge.move instanceof EntityMove)) {
        // We don't want to visit field edges as we're looking for indirect paths
        continue;
      }

      // A huge win for performance, is when you do less work :D
      // We can ignore an edge that has already been visited with the same key fields / requirements.
      // The way entity-move edges are created, where every graph points to every other graph:
      //  Graph A: User @key(id) @key(name)
      //  Graph B: User @key(id)
      //  Edges in a merged graph:
      //    - User/A @key(id) -> User/B
      //    - User/B @key(id) -> User/A
      //    - User/B @key(name) -> User/A
      // Allows us to ignore an edge with the same key fields.
      // That's because in some other path, we will or already have checked the other edge.
      if (
        !!edge.requirement &&
        visitedKeyFields.some((f) => f.equals(edge.requirement!))
      ) {
        continue;
      }

      // TODO: check if the requirement can be satisfied by the current path
      if (edge.requirement) {
        //
      }

      const newPath = path.advance(edge);

      const directPaths = findDirectPaths(graph, newPath, step);

      if (directPaths.length) {
        nextPaths.push(...directPaths);
        continue;
      }

      queue.push([
        concatIfNotExistsString(visitedGraphs, edge.tail.subgraphId),
        !!edge.requirement
          ? concatIfNotExistsFields(visitedKeyFields, edge.requirement)
          : visitedKeyFields,
        newPath,
      ]);
    }
  }

  return nextPaths;
}

function concatIfNotExistsString(list: string[], item: string): string[] {
  if (list.includes(item)) {
    return list;
  }

  return list.concat(item);
}

function concatIfNotExistsFields(
  list: Selection[],
  item: Selection,
): Selection[] {
  if (list.some((f) => f.equals(item))) {
    return list;
  }

  return list.concat(item);
}

class OperationPath {
  constructor(
    public rootNode: Node,
    public edges: Edge[] = [],
  ) {}

  advance(edge: Edge) {
    return new OperationPath(this.rootNode, this.edges.concat(edge));
  }

  tail(): Node {
    if (this.edges.length) {
      return this.edges[this.edges.length - 1].tail;
    }

    return this.rootNode;
  }
}
