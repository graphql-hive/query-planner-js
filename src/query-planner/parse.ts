import {
  Kind,
  ObjectTypeDefinitionNode,
  parse,
  specifiedScalarTypes,
} from "graphql";
import { invariant, isListTypeNode, resolveTypeNodeName } from "./utils";
import { Edge, FieldMove, Graph, Node } from "./graph";
import {
  ObjectType,
  ObjectTypeField,
  Subgraph,
  Supergraph,
  TypeKind,
} from "./schema";
import {
  JoinField,
  JoinType,
  parseJoinField,
  parseJoinType,
} from "./join-directives";
import { SelectionResolver } from "./selection-resolver";

export function parseSupergraph(sdl: string) {
  const supergraph = new Supergraph();
  const supergraphDocument = parse(sdl);

  for (const def of supergraphDocument.definitions) {
    if (def.kind === Kind.OBJECT_TYPE_DEFINITION) {
      const objectTypes = parseObjectType(def);

      for (const objectType of objectTypes) {
        for (const joinType of objectType.join) {
          const subgraph = ensureSubgraph(supergraph, joinType.graph);
          subgraph.types.set(objectType.name, objectType);
          if (joinType.resolvable && typeof joinType.key === "string") {
            subgraph.entityTypes.add(objectType.name);
          }
        }
      }
    }
  }

  const graphs: Graph[] = [];
  const entities = new Map<
    string,
    Array<{
      key: string;
      graphId: string;
      typeName: string;
    }>
  >();
  const selectionResolvers = new Map<string, SelectionResolver>();

  for (const [graphId, subgraph] of supergraph.subgraphs) {
    graphs.push(buildGraphFromSubgraph(subgraph));
    selectionResolvers.set(graphId, new SelectionResolver(subgraph));

    for (const [typeName, typeState] of subgraph.types) {
      if (!Array.isArray(entities.get(typeName))) {
        entities.set(typeName, []);
      }

      for (const joinType of typeState.join) {
        if (!joinType.resolvable || typeof joinType.key !== "string") {
          continue;
        }
        entities.get(typeName)!.push({
          key: joinType.key,
          graphId,
          typeName: typeState.name,
        });
      }
    }
  }

  const mergedGraph = new Graph("private");

  for (const graph of graphs) {
    mergedGraph.copyFrom(graph);
  }

  mergedGraph.joinByKeys(entities, selectionResolvers);

  return mergedGraph;
}

function parseObjectType(def: ObjectTypeDefinitionNode): ObjectType[] {
  const joinTypes: JoinType[] = [];
  const fields = def.fields;
  const graphIds = new Set<string>();

  for (const directiveNode of def.directives ?? []) {
    if (directiveNode.name.value === "join__type") {
      const joinType = parseJoinType(directiveNode);
      graphIds.add(joinType.graph);
      joinTypes.push(joinType);
    }
  }

  const objectTypePerGraph = new Map<string, ObjectType>();

  for (const graphId of graphIds) {
    const objectType = new ObjectType(
      def.name.value,
      joinTypes.filter((joinType) => joinType.graph === graphId),
    );
    objectTypePerGraph.set(graphId, objectType);
  }

  if (!fields?.length) {
    return Array.from(objectTypePerGraph.values());
  }

  for (const field of fields) {
    const isList = isListTypeNode(field.type);
    if (!field.directives?.length) {
      // it means that the field belongs to all graphs defining the object type

      for (const [graphId, objectType] of objectTypePerGraph) {
        objectType.addField(
          new ObjectTypeField(
            field.name.value,
            resolveTypeNodeName(field.type),
            isList,
            // TODO: turn it into object...
            new JoinField(graphId, null, null, null, false, null, false),
          ),
        );
      }

      continue;
    }

    const joinFields: JoinField[] = [];

    for (const directiveNode of field.directives ?? []) {
      if (directiveNode.name.value === "join__field") {
        joinFields.push(parseJoinField(directiveNode));
      }
    }

    for (const joinField of joinFields) {
      invariant(
        typeof joinField.graph === "string",
        "No support for fields provided by an interface object yet",
      );
      const objectType = objectTypePerGraph.get(joinField.graph);
      invariant(
        !!objectType,
        "No object type found for graph " + joinField.graph,
      );

      objectType.addField(
        new ObjectTypeField(
          field.name.value,
          resolveTypeNodeName(field.type),
          isList,
          joinField,
        ),
      );
    }
  }

  return Array.from(objectTypePerGraph.values());
}

function buildGraphFromSubgraph(subgraph: Subgraph): Graph {
  const graph = new Graph(subgraph.graphId);

  const queryType = subgraph.types.get("Query");

  // Starts from root
  if (queryType) {
    createNodesAndEdgesForObjectType(graph, subgraph, queryType);
  }

  // Adds entity types
  subgraph.entityTypes.forEach((typeName) => {
    const objectType = subgraph.types.get(typeName);
    invariant(!!objectType, "Type not found: " + typeName);

    createNodesAndEdgesForObjectType(graph, subgraph, objectType);
  });

  return graph;
}

function createNodesAndEdgesForType(
  graph: Graph,
  subgraph: Subgraph,
  typeName: string,
) {
  if (specifiedScalarTypes.some((t) => t.name === typeName)) {
    return createNodeForScalarType(graph, subgraph, typeName);
  }

  const typeState = subgraph.types.get(typeName);
  invariant(!!typeState, "Type not found: " + typeName);
  return createNodesAndEdgesForObjectType(graph, subgraph, typeState);
}

function createNodesAndEdgesForObjectType(
  graph: Graph,
  subgraph: Subgraph,
  objectType: ObjectType,
) {
  const existing = graph.ensureNonOrSingleNode(objectType.name);
  if (existing) {
    return existing;
  }

  const head = graph.createTypeNode(objectType.name, TypeKind.OBJECT_TYPE);

  for (const field of objectType.fields) {
    if (field.join.external) {
      continue;
    }
    createEdgeForObjectTypeField(graph, subgraph, head, field);
  }

  return head;
}

function createEdgeForObjectTypeField(
  graph: Graph,
  subgraph: Subgraph,
  head: Node,
  field: ObjectTypeField,
) {
  const tail = createNodesAndEdgesForType(graph, subgraph, field.type);

  if (!tail) {
    throw new Error(
      `Failed to create Node for ${field.type} in subgraph ${graph.id}`,
    );
  }

  return graph.addEdge(
    new Edge(
      head,
      tail,
      new FieldMove(field.name, head.typeName, head.typeKind, field.isList),
      null,
    ),
  );
}

function createNodeForScalarType(
  graph: Graph,
  subgraph: Subgraph,
  typeName: string,
) {
  const existing = graph.ensureNonOrSingleNode(typeName);
  if (existing) {
    return existing;
  }

  return graph.createTypeNode(typeName, TypeKind.SCALAR_TYPE);
}

function ensureSubgraph(supergraph: Supergraph, graphId: string) {
  const existing = supergraph.subgraphs.get(graphId);

  if (existing) {
    return existing;
  }

  const subgraph = new Subgraph(graphId);
  supergraph.subgraphs.set(graphId, subgraph);
  return subgraph;
}
