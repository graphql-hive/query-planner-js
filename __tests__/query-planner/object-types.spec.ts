import { describe, expect, test } from "vitest";
import { assertCompositionSuccess, graphql } from "../shared/testkit";
import { OperationTypeNode, parse } from "graphql";
import { composeServices } from "../../src/compose.js";
import { parseSupergraph } from "../../src/query-planner/parse";
import { OperationPath, walkQuery } from "../../src/query-planner/walker";
import {
  generateQueryPlan,
  prettyPrintQueryPlan,
} from "../../src/query-planner/plan";

function printPath(path: OperationPath) {
  console.log(path.edges.map((edge) => edge.toString()).join(" -> "));
  let i = 0;
  for (const requiredPathsOfEdge of path.requiredPathsForEdges) {
    if (requiredPathsOfEdge.length) {
      console.log(" edge " + path.edges[i++].toString() + "depends on: ");
      for (const requiredPath of requiredPathsOfEdge) {
        console.log(
          "  " + requiredPath.edges.map((edge) => edge.toString()).join(" -> "),
        );
      }
    }
  }
}

describe("Object Types Entities", () => {
  test("single key field", () => {
    const result = composeServices([
      {
        name: "a",
        typeDefs: graphql`
          extend schema
            @link(
              url: "https://specs.apollo.dev/federation/v2.3"
              import: ["@key", "@shareable"]
            )

          type User @key(fields: "id") {
            id: ID!
            name: String @shareable
            age: Int
          }
        `,
      },
      {
        name: "b",
        typeDefs: graphql`
          extend schema
            @link(
              url: "https://specs.apollo.dev/federation/v2.3"
              import: ["@key", "@shareable"]
            )

          type User @key(fields: "id") {
            id: ID!
            name: String @shareable
          }

          type Query {
            users: [User] @shareable
          }
        `,
      },
    ]);

    assertCompositionSuccess(result);

    const supergraph = parseSupergraph(result.supergraphSdl);
    const path = walkQuery(supergraph, OperationTypeNode.QUERY, [
      {
        kind: "Field",
        name: "users",
      },
      {
        kind: "Field",
        name: "age",
      },
    ]);
    expect(path).not.toBeNull();

    const plan = generateQueryPlan(path!);
    expect(plan).toMatchInlineSnapshot(`
      {
        "kind": "QueryPlan",
        "node": {
          "kind": "Sequence",
          "nodes": [
            {
              "kind": "Fetch",
              "operation": "{
      users {
        __typename

        id
        }
      }
      ",
              "operationKind": "query",
              "serviceName": "B",
              "variableUsages": [],
            },
            {
              "kind": "Flatten",
              "node": {
                "kind": "Fetch",
                "operation": "query ($representations : [_Any!]!  ){
        _entities(representations: $representations) {
          ... on User {
            age
          }
        }
      }",
                "operationKind": "query",
                "requires": {
                  "kind": "fragment",
                  "selectionSet": [
                    {
                      "fieldName": "id",
                      "kind": "field",
                      "selectionSet": null,
                      "typeName": "User",
                    },
                  ],
                  "typeName": "User",
                },
                "serviceName": "A",
                "variableUsages": [],
              },
              "path": [
                "users",
                "@",
              ],
            },
          ],
        },
      }
    `);
  });
  test.only("complex-entity-call", () => {
    const result = composeServices([
      {
        name: "link",
        typeDefs: graphql`
          extend schema
            @link(
              url: "https://specs.apollo.dev/federation/v2.0"
              import: ["@key"]
            )

          type Product @key(fields: "id") @key(fields: "id pid") {
            id: String!
            pid: String!
          }
        `,
      },
      {
        name: "list",
        typeDefs: graphql`
          extend schema
            @link(
              url: "https://specs.apollo.dev/federation/v2.0"
              import: ["@key", "@shareable"]
            )

          type ProductList @key(fields: "products{id pid}") {
            products: [Product!]!
            first: Product @shareable
            selected: Product @shareable
          }

          type Product @key(fields: "id pid") {
            id: String!
            pid: String
          }
        `,
      },
      {
        name: "price",
        typeDefs: graphql`
          extend schema
            @link(
              url: "https://specs.apollo.dev/federation/v2.0"
              import: ["@key", "@shareable"]
            )

          type ProductList
            @key(fields: "products{id pid category{id tag}} selected{id}") {
            products: [Product!]!
            first: Product @shareable
            selected: Product @shareable
          }

          type Product @key(fields: "id pid category{id tag}") {
            id: String!
            price: Price
            pid: String
            category: Category
          }

          type Category @key(fields: "id tag") {
            id: String!
            tag: String
          }

          type Price {
            price: Float!
          }
        `,
      },
      {
        name: "products",
        typeDefs: graphql`
          extend schema
            @link(
              url: "https://specs.apollo.dev/federation/v2.0"
              import: ["@key", "@external", "@extends", "@shareable"]
            )

          type Query {
            topProducts: ProductList!
          }

          type ProductList @key(fields: "products{id}") {
            products: [Product!]!
          }

          type Product @extends @key(fields: "id") {
            id: String! @external
            category: Category @shareable
          }

          type Category @key(fields: "id") {
            mainProduct: Product! @shareable
            id: String!
            tag: String @shareable
          }
        `,
      },
    ]);
    assertCompositionSuccess(result);
    const supergraph = parseSupergraph(result.supergraphSdl);
    const path = walkQuery(supergraph, OperationTypeNode.QUERY, [
      {
        kind: "Field",
        name: "topProducts",
      },
      {
        kind: "Field",
        name: "products",
      },
      {
        kind: "Field",
        name: "price",
      },
      {
        kind: "Field",
        name: "price",
      },
    ]);
    expect(path).not.toBeNull();

    printPath(path!);

    const plan = generateQueryPlan(path!);
    expect(plan).toMatchInlineSnapshot(`
      {
        "kind": "QueryPlan",
        "node": {
          "kind": "Sequence",
          "nodes": [
            {
              "kind": "Fetch",
              "operation": "{
      topProducts {
        __typename

        products {
          __typename

          category {
            id
            tag
          }
          id
          pid
          }
        }
      }
      ",
              "operationKind": "query",
              "serviceName": "PRODUCTS",
              "variableUsages": [],
            },
            {
              "kind": "Flatten",
              "node": {
                "kind": "Fetch",
                "operation": "query ($representations : [_Any!]!  ){
        _entities(representations: $representations) {
          ... on Product {
            price
          }
        }
      }",
                "operationKind": "query",
                "requires": {
                  "kind": "fragment",
                  "selectionSet": [
                    {
                      "fieldName": "category",
                      "kind": "field",
                      "selectionSet": [
                        {
                          "fieldName": "id",
                          "kind": "field",
                          "selectionSet": null,
                          "typeName": "Category",
                        },
                        {
                          "fieldName": "tag",
                          "kind": "field",
                          "selectionSet": null,
                          "typeName": "Category",
                        },
                      ],
                      "typeName": "Product",
                    },
                    {
                      "fieldName": "id",
                      "kind": "field",
                      "selectionSet": null,
                      "typeName": "Product",
                    },
                    {
                      "fieldName": "pid",
                      "kind": "field",
                      "selectionSet": null,
                      "typeName": "Product",
                    },
                  ],
                  "typeName": "Product",
                },
                "serviceName": "PRICE",
                "variableUsages": [],
              },
              "path": [
                "topProducts",
                "products",
                "@",
              ],
            },
          ],
        },
      }
    `);

    expect(prettyPrintQueryPlan(plan)).toMatchInlineSnapshot(`
      "QueryPlan {
        Sequence {
          Fetch(service: "PRODUCTS") {
            {
              topProducts{
                __typename
                products{
                  __typename
                  category{
                    id
                    tag
                  }
                  id
                  pid
                }
              }
            }
          },
          Flatten(path: "topProducts.products.@") {
            Fetch(service: "PRICE") {
              {
                ... on Product {
                  category {
                    id
                    tag
                  }
                  id
                  pid
                }
              } =>
              {
                ... on Product {
                  price
                }
              }
            }
          }
        }
      }"
    `);
  });
});
