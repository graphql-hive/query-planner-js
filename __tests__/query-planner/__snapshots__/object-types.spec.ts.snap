// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Object Types Entities > complex-entity-call 2`] = `
"QueryPlan {
  Sequence {
    Fetch(service: "PRODUCTS") {
      {
        topProducts{
          __typename
          products{
            __typename
            category{
              id
              tag
            }
            id
            pid
          }
        }
      }
    },
    Flatten(path: "topProducts.products.@") {
      Fetch(service: "PRICE") {
        {
          ... on Product {
            category {
              id
              tag
            }
            id
            pid
          }
        } =>
        {
          ... on Product {
            price
          }
        }
      }
    }
  }
}"
`;
